### DISCUSSION TITLE
Possible to change data fetching method based on branch?

### DISCUSSION THREAD

**Discussion by dungle-scrubs** [vercel/community#1075] [Help]
Is there a pattern to set the data fetching method for a page, depending on its branch?

For example, in `production`, the page should use `getStaticProps` (with or without `getStaticPaths`).

However, if the site is using a CMS and your client wants to see their changes made on refresh, you'd need to use `getServerSideProps`, else run a full deploy. 

Assuming that it's setup such that both `getStaticProps` and `getServerSide` props can both use the exact same code, is there a way to set it such that when the branch is `staging` and deployed to Vercel, the pages use `getServerSideProps`? 

Thanks

**Comment by amyegan** [comment]
Hi @dungle-scrubs! If you can tell me more about why you want static rendering in production but server-side rendering in preview, I can help you find an appropriate solution. 

[Incremental Static Regeneration](https://vercel.com/docs/concepts/incremental-static-regeneration/overview#) might be a good solution for you. We have some ISR demos and templates if you want to check it out:
- [GitHub Reactions](https://reactions-demo.vercel.app/) demo
- [Static Tweets](https://static-tweet.vercel.app/) demo
- [On-Demand ISR](https://vercel.com/templates/next.js/on-demand-incremental-static-regeneration) template
- [ISR Blog with Next.js and WordPress](https://vercel.com/templates/next.js/isr-blog-nextjs-wordpress) template

**Comment by dungle-scrubs** [comment]
Hey @amyegan , sure thing. My use case is the following:

I'm building a Next.js site for a client where in the end, most pages will be statically generated for speed. We're using Prismic as a CMS and also data from an API that they are hosting. 

When they make changes to either Prismic or their API, they need to do a full re-deploy to see those changes. With Prismic it's less of an issue because they offer preview functionality so they don't need to do a full deploy in that case.

However, with their own API, which is being updated frequently during this development phase, each time I provide them with a preview deploy on our `staging` branch, the pages are statically generated, so they cannot easily see their changes in real time.

What I'm doing currently is setting up all data fetching using `getServerSide` props, even in production. So when we all want to gauge performance, I switch over to `getStaticProps` along with `getStaticPaths`.  Changing back and forth isn't very ideal. 

So I'm wondering if there's a way to "toggle" between the two methods. 

Of course, I might be completely missing something in our workflow where none of this is necessary.  I'm looking over the links you provided and don't fully understand ISR yet, but will give it a more thorough read soon.

Thanks for the help

**Reply by amyegan** [reply]
[On-Demand Revalidation](https://vercel.com/docs/concepts/incremental-static-regeneration/quickstart#on-demand-revalidation) would be perfect for this situation. It essentially let's you rebuild just the updated content as-needed (or after a set amount of time is using background revalidation). You could use it for preview deployments and skip it in production to keep the production site completely static with the normal `getStaticProps` function. But you could also use on-demand ISR in production to avoid a full site redeploy for content updates.

The [Next.js ISR page](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration) explains more of the implementation details. Prismic also shared an article about this at [prismic.io/blog/nextjs-sites-on-demand-isr](https://prismic.io/blog/nextjs-sites-on-demand-isr). Please let me know if you have any questions about it!
